source/agora/test/Base.d

```diff
@@ -907,6 +914,170 @@ public class TestAPIManager
+            format!"[%s:%s] No blocks in first client!"(file, line));
+
+        // Get spendables from last block
+        auto spendables = all_blocks[$ - 1].spendable().array;
+
+        // Ensure at least one tx will be taken
+        auto tx_count = max(1, this.test_conf.txs_to_nominate);
+
+        // Show the last block if not enough spendables
+        assert(spendables.length >= tx_count,
+            format!"[%s:%s] Less than %s spendables in block:\n%s"
+                (file, line, tx_count, prettify(all_blocks[$ - 1])));
+
+        // Send transactions to the first client
+        spendables.takeExactly(tx_count)
+            .map!(txb => txb.sign())
+            .each!(tx => first_client.putTransaction(tx));
+
+        // Get preimage distance from enrollment to this next block
+        ushort distance = cast(ushort) (target_height - enrolled_height -1);
+        assert(distance < GenesisValidatorCycle && distance >= 0,
+            format!"[%s:%s] Expected distance between 0 and %s not %s"
+                (file, line, GenesisValidatorCycle - 1, distance));
+
+        // Check block is at target height for the participating clients
+        expectBlock(client_idxs, target_height,
+            all_blocks[enrolled_height].header);
+    }
+
+    /***************************************************************************
+
+        Enroll validator
+
+        This is a helper function to enroll a validator and wait till
+            other validators have the enroll on their pool
+
+        Params:
+            client_idx = the index of the client to enroll
+            client_idxs = client indices for the participating validators
+
+    ***************************************************************************/
+
+    void enroll (size_t client_idx)
+    {
+        enroll(iota(0, GenesisValidators), client_idx);
+    }
+
+    /// Ditto
+    void enroll (Idxs)(Idxs client_idxs, size_t client_idx,
+        string file = __FILE__, size_t line = __LINE__)
+    {
+        static assert (isInputRange!Idxs);
+
+        Enrollment enroll = this.clients[client_idx].createEnrollmentData();
+        clients[client_idx].enrollValidator(enroll);
+        client_idxs.each!(idx =>
+            retryFor(this.clients[idx].getEnrollment(enroll.utxo_key) == enroll,
+                5.seconds,
+                format!"[%s:%s] Client #%s enrollment not in pool of client #%s"
+                (file, line, client_idx, idx)));
+    }
+
+    /// Returns: true if all the nodes contain the same blocks
+    public bool containSameBlocks (Idxs)(Idxs client_idxs, size_t height)
+    {
+        static assert (isInputRange!Idxs);
+
+        if (client_idxs.any!(idx =>
+            this.clients[idx].getBlockHeight() != height))
+            return false;
+
+        auto first_blocks = this.clients[client_idxs.front].getAllBlocks();
+
+        return !(client_idxs.drop(1).any!(idx =>
+            this.clients[idx].getAllBlocks() != first_blocks));
```

Comment:
Why use ! and `any` when you can just use `all`?

```suggestion
        return client_idxs.drop(1).all!(idx =>
            this.clients[idx].getAllBlocks() == first_blocks);
```

Reply:
It will fail faster this way.

Reply:
?

`all` will short-circuit too.

Reply:
Actually I guess it will still fail at first not matching so ignore my previous comment.

Reply:
```D
import std;

bool func (int a) { writeln("Value: ", a); return a == 2; }
void main ()
{
 	const values = [ 1, 2, 3, 4 ];
    writeln("ALL:");
    bool b = values.all!func;
    writeln("ANY:");
    bool c = values.any!func;    
}import std;

bool func (int a) { writeln("Value: ", a); return a == 2; }
void main ()
{
 	const values = [ 1, 2, 3, 4 ];
    writeln("ALL:");
    bool b = values.all!func;
    writeln("ANY:");
    bool c = values.any!func;    
}
```

```
ALL:
Value: 1
ANY:
Value: 1
Value: 2
```

Reply:
Updated to `all` :-)